images/base/files/kind/bin/undo-mount-hacks.sh

```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
+
+# Copyright 2023 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+# don't run on cgroups v2
+if [[ -f "/sys/fs/cgroup/cgroup.controllers" ]]; then
+    exit 0
+fi
+
+# on cgroups v1 we want to undo bind mounting over this to hide misc
+# see entrypoint
+umount /proc/cgroups
```

Comment:
don't this run after the entrypoint?

Reply:
yes, the idea is we don't want to undo this until _after_ systemd read the file and setup cgroups. we don't terribly need to rush here and may not need to do it at all, but conceivably something like cadvisor could desire to read the real details and it's just confusing to have a fake here because the contents are dynamic, they include the number of active groups for the controller.

Reply:
https://man7.org/linux/man-pages/man7/cgroups.7.html

```
   /proc files
       /proc/cgroups (since Linux 2.6.24)
              This file contains information about the controllers that
              are compiled into the kernel.  An example of the contents
              of this file (reformatted for readability) is the
              following:

                  #subsys_name    hierarchy      num_cgroups    enabled
                  cpuset          4              1              1
                  cpu             8              1              1
                  cpuacct         8              1              1
                  blkio           6              1              1
                  memory          3              1              1
                  devices         10             84             1
                  freezer         7              1              1
                  net_cls         9              1              1
                  perf_event      5              1              1
                  net_prio        9              1              1
                  hugetlb         0              1              0
                  pids            2              1              1

              The fields in this file are, from left to right:

              1. The name of the controller.

              2. The unique ID of the cgroup hierarchy on which this
                 controller is mounted.  If multiple cgroups v1
                 controllers are bound to the same hierarchy, then each
                 will show the same hierarchy ID in this field.  The
                 value in this field will be 0 if:

                   a) the controller is not mounted on a cgroups v1
                      hierarchy;

                   b) the controller is bound to the cgroups v2 single
                      unified hierarchy; or

                   c) the controller is disabled (see below).

              3. The number of control groups in this hierarchy using
                 this controller.

              4. This field contains the value 1 if this controller is
                 enabled, or 0 if it has been disabled (via the
                 cgroup_disable kernel command-line boot parameter).
```

We're deleting the `misc` line and making a fixed copy of this what systemd sees when probing for available controllers.

Reply:
systemd never retries, right?

Reply:
correct. from systemd POV everything succeeded and cgroup controller mounting is a one-time init thing.

Reply:
this is all because, if we controlled the host kernel etc. we'd either just have misc aware all the way down or disable the controller in the kernel. it doesn't make sense from systemd POV to have a flag to disable controllers because:

1) in a non containerized environment this could already be done in the kernel.
2) in a containerized environment systemd is not supposed to be in a privileged container and cgroups should be fully setup already.

and changing this isn't really worth it, because even if you could try to make this sound reasonable, it's all deprecated cgroups v1 anyhow, systemd and the rest of the ecosystem will drop that in the next year or so and v2 doesn't need any of this.